{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hamming_distance(p: str, q: str) -> int:\n",
    "    hdist = 0\n",
    "    for i in range(len(p)):\n",
    "        if p[i] != q[i]:\n",
    "            hdist += 1\n",
    "    return hdist\n",
    "\n",
    "def neighbors(s: str, d: int) -> list[str]:\n",
    "    nucleotides = {'A', 'T', 'C', 'G'}\n",
    "    neighborhood = set()\n",
    "    if d == 0:\n",
    "        neighborhood.add(s)\n",
    "        return neighborhood\n",
    "    if len(s) == 1:\n",
    "        neighborhood.update(['A', 'T', 'C', 'G'])\n",
    "        return neighborhood\n",
    "    suffixNeighbors = neighbors(s[1:], d)\n",
    "    for str in suffixNeighbors:\n",
    "        if hamming_distance(s[1:], str) < d:\n",
    "            for nuc in nucleotides:\n",
    "                neighborhood.add(nuc+str)\n",
    "        else:\n",
    "            neighborhood.add(s[0]+str)\n",
    "    return neighborhood\n",
    "\n",
    "def motif_enumeration(dna: list[str], k: int, d: int) -> list[str]:\n",
    "    patterns = []\n",
    "    for pattern in dna:\n",
    "        for i in range(len(pattern) - k + 1):\n",
    "            kmer = pattern[i:i+k]\n",
    "            kmerNeighbors = neighbors(kmer, d)\n",
    "            for neighbor in kmerNeighbors:\n",
    "                count_DNA_neighbor_is_in = 0\n",
    "                for dnaString in dna:\n",
    "                    found_in_DNA_string = False\n",
    "                    for j in range(len(dnaString) - k + 1):\n",
    "                        kmer1 = dnaString[j:j+k]\n",
    "                        if hamming_distance(kmer1, neighbor) <= d:\n",
    "                            found_in_DNA_string = True\n",
    "                            break\n",
    "                    if found_in_DNA_string:\n",
    "                        count_DNA_neighbor_is_in += 1\n",
    "                if count_DNA_neighbor_is_in == len(dna):\n",
    "                    if neighbor not in patterns:\n",
    "                        patterns.append(neighbor)\n",
    "    return patterns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hamming_distance(p: str, q: str) -> int:\n",
    "    hdist = 0\n",
    "    for i in range(len(p)):\n",
    "        if p[i] != q[i]:\n",
    "            hdist += 1\n",
    "    return hdist\n",
    "\n",
    "def DistanceBetweenPatternAndStrings (pattern: str, dna: list[str]) -> int:\n",
    "    k = len(pattern)\n",
    "    distance = 0\n",
    "    for dnaString in dna:\n",
    "        hammingDist = float('inf')\n",
    "        for i in range(len(dnaString) - k + 1):\n",
    "            kmer = dnaString[i:i+k]\n",
    "            if hammingDist > hamming_distance(pattern, kmer):\n",
    "                hammingDist = hamming_distance(pattern, kmer)\n",
    "        distance += hammingDist   \n",
    "    return distance\n",
    "\n",
    "def AllStrings(k: int) -> list[str]:\n",
    "    nucleotides = ['A', 'T', 'C', 'G']\n",
    "    if k == 1:\n",
    "        return nucleotides\n",
    "    smaller_strings = AllStrings(k-1)\n",
    "    kmers = []\n",
    "    for nuc in nucleotides:\n",
    "        for string in smaller_strings:\n",
    "            kmers.append(nuc + string)\n",
    "    return kmers\n",
    "\n",
    "def median_string(dna: list[str], k: int) -> str:\n",
    "    distance = float('inf')\n",
    "    patterns = AllStrings(k)\n",
    "    for i in range(len(patterns)):\n",
    "        pattern = patterns[i]\n",
    "        if distance > DistanceBetweenPatternAndStrings(pattern, dna):\n",
    "            distance = DistanceBetweenPatternAndStrings(pattern, dna)\n",
    "            median = pattern\n",
    "    return median"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def profile_most_probable_kmer(text: str, k: int,\n",
    "                               profile: list[dict[str, float]]) -> str:\n",
    "    \"\"\"Identifies the most probable k-mer according to a given profile matrix.\n",
    "\n",
    "    The profile matrix is represented as a list of columns, where the i-th element is a map\n",
    "    whose keys are strings (\"A\", \"C\", \"G\", and \"T\") and whose values represent the probability\n",
    "    associated with this symbol in the i-th column of the profile matrix.\n",
    "    \"\"\"\n",
    "    probs = {}\n",
    "    for i in range(len(text) - k + 1):\n",
    "        prob = 1\n",
    "        kmer = text[i:i + k]\n",
    "        for i in range(k):\n",
    "            prob *= profile[i][kmer[i]]\n",
    "        probs[kmer] = prob\n",
    "    maxProb = max(probs.values())\n",
    "    for kmer in probs.keys():\n",
    "        if probs[kmer] == maxProb:\n",
    "            return kmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_profile(motifs: list[str]) -> list[dict[str, float]]:\n",
    "    t = len(motifs)\n",
    "    k = len(motifs[0])\n",
    "    profile = [{\"A\":1, \"T\":1, \"C\":1, \"G\":1} for _ in range(k)]\n",
    "    for i in range(k):\n",
    "        for motif in motifs:\n",
    "            profile[i][motif[i]] += 1/t\n",
    "    return profile\n",
    "\n",
    "def score(motifs: list[str], profile: list[dict[str, float]]) -> int:\n",
    "    consensus = \"\"\n",
    "    count = 0\n",
    "    for i in range(len(profile)):\n",
    "        column = profile[i]\n",
    "        m = max(column.values())\n",
    "        for key in column.keys():\n",
    "            if len(consensus) == i:\n",
    "                if column[key] == m:\n",
    "                    consensus += key\n",
    "    for motif in motifs:\n",
    "        for i in range(len(motif)):\n",
    "            if motif[i] != consensus[i]:\n",
    "                count += 1\n",
    "    return count\n",
    "\n",
    "def profile_most_probable_kmer(text: str, k: int, profile: list[dict[str, float]]) -> str:\n",
    "    probs = {}\n",
    "    for i in range(len(text) - k + 1):\n",
    "        prob = 1\n",
    "        kmer = text[i:i + k]\n",
    "        for i in range(k):\n",
    "            prob *= profile[i][kmer[i]]\n",
    "        probs[kmer] = prob\n",
    "    maxProb = max(probs.values())\n",
    "    for kmer in probs.keys():\n",
    "        if probs[kmer] == maxProb:\n",
    "            return kmer\n",
    "\n",
    "\n",
    "# Please do not remove package declarations because these are used by the autograder.\n",
    "# Insert your greedy_motif_search function here, along with any subroutines you need\n",
    "def greedy_motif_search_pseudocounts(dna: list[str], k: int, t: int) -> list[str]:\n",
    "    \"\"\"Implements the GreedyMotifSearch algorithm.\"\"\"\n",
    "    best_motifs = []\n",
    "    for string in dna:\n",
    "        best_motifs.append(string[:k])\n",
    "    best_score = score(best_motifs, generate_profile(best_motifs))\n",
    "    for i in range(len(dna[0]) - k + 1):\n",
    "        motif1 = dna[0][i:i+k]\n",
    "        motifs = [motif1]\n",
    "        profile = generate_profile(motifs)\n",
    "        for i in range(1, t):\n",
    "            motifs.append(profile_most_probable_kmer(dna[i], k, profile))\n",
    "            profile = generate_profile(motifs)\n",
    "        new_score = score(motifs, profile)\n",
    "        if new_score < best_score:\n",
    "            best_score = new_score\n",
    "            best_motifs = motifs\n",
    "    return best_motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "def generate_profile(motifs: list[str]) -> list[dict[str, float]]:\n",
    "    t = len(motifs)\n",
    "    k = len(motifs[0])\n",
    "    profile = [{\"A\":1, \"T\":1, \"C\":1, \"G\":1} for _ in range(k)]\n",
    "    for i in range(k):\n",
    "        for motif in motifs:\n",
    "            profile[i][motif[i]] += 1/t\n",
    "    return profile\n",
    "\n",
    "def score(motifs: list[str], profile: list[dict[str, float]]) -> int:\n",
    "    consensus = \"\"\n",
    "    count = 0\n",
    "    for i in range(len(profile)):\n",
    "        column = profile[i]\n",
    "        m = max(column.values())\n",
    "        for key in column.keys():\n",
    "            if len(consensus) == i:\n",
    "                if column[key] == m:\n",
    "                    consensus += key\n",
    "    for motif in motifs:\n",
    "        for i in range(len(motif)):\n",
    "            if motif[i] != consensus[i]:\n",
    "                count += 1\n",
    "    return count\n",
    "\n",
    "def profile_most_probable_kmer(text: str, k: int, profile: list[dict[str, float]]) -> str:\n",
    "    probs = {}\n",
    "    for i in range(len(text) - k + 1):\n",
    "        prob = 1\n",
    "        kmer = text[i:i + k]\n",
    "        for i in range(k):\n",
    "            prob *= profile[i][kmer[i]]\n",
    "        probs[kmer] = prob\n",
    "    maxProb = max(probs.values())\n",
    "    for kmer in probs.keys():\n",
    "        if probs[kmer] == maxProb:\n",
    "            return kmer\n",
    "\n",
    "def randomized_motif_search(dna: list[str], k: int, t: int) -> list[str]:\n",
    "    best_motifs = []\n",
    "    for string in dna:\n",
    "        randomStart = random.randint(0, len(string) - k)\n",
    "        randomKmer = string[randomStart:randomStart + k]\n",
    "        best_motifs.append(randomKmer)\n",
    "    best_score = score(best_motifs, generate_profile(best_motifs))\n",
    "    while True:\n",
    "        profile = generate_profile(best_motifs)\n",
    "        motifs = []\n",
    "        for i in range(t):\n",
    "            motifs.append(profile_most_probable_kmer(dna[i], k, profile))\n",
    "            profile = generate_profile(motifs)\n",
    "        new_score = score(motifs, profile)\n",
    "        if new_score < best_score:\n",
    "            best_score = new_score\n",
    "            best_motifs = motifs\n",
    "        else:\n",
    "            return best_motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys \n",
    "import random\n",
    "\n",
    "def generate_profile(motifs: list[str]) -> list[dict[str, float]]:\n",
    "    t = len(motifs)\n",
    "    k = len(motifs[0])\n",
    "    profile = [{\"A\":1, \"T\":1, \"C\":1, \"G\":1} for _ in range(k)]\n",
    "    for i in range(k):\n",
    "        for motif in motifs:\n",
    "            profile[i][motif[i]] += 1/t\n",
    "    return profile\n",
    "\n",
    "def score(motifs: list[str], profile: list[dict[str, float]]) -> int:\n",
    "    consensus = \"\"\n",
    "    count = 0\n",
    "    for i in range(len(profile)):\n",
    "        column = profile[i]\n",
    "        m = max(column.values())\n",
    "        for key in column.keys():\n",
    "            if len(consensus) == i:\n",
    "                if column[key] == m:\n",
    "                    consensus += key\n",
    "    for motif in motifs:\n",
    "        for i in range(len(motif)):\n",
    "            if motif[i] != consensus[i]:\n",
    "                count += 1\n",
    "    return count\n",
    "\n",
    "def profile_weighted_random_kmer(text: str, k: int, profile: list[dict[str, float]]) -> str:\n",
    "    probs = {}\n",
    "    for i in range(len(text) - k + 1):\n",
    "        prob = 1\n",
    "        kmer = text[i:i + k]\n",
    "        for i in range(k):\n",
    "            prob *= profile[i][kmer[i]]\n",
    "        probs[kmer] = prob\n",
    "    selectedKmer = random.choices(list(probs.keys()), list(probs.values()), k=1)\n",
    "    return selectedKmer[0]\n",
    "\n",
    "def gibbs_sampler(dna: list[str], k: int, t: int, n: int) -> list[str]:\n",
    "    best_motifs = []\n",
    "    for string in dna:\n",
    "        randomStart = random.randint(0, len(string) - k)\n",
    "        randomKmer = string[randomStart:randomStart + k]\n",
    "        best_motifs.append(randomKmer)\n",
    "    best_score = float('inf')\n",
    "    for _ in range(n):\n",
    "        i = random.randint(0,t-1)\n",
    "        motifs = best_motifs.copy()\n",
    "        motifs.pop(i)\n",
    "        profile = generate_profile(motifs)\n",
    "        motifi = profile_weighted_random_kmer(dna[i], k, profile)\n",
    "        motifs.insert(i, motifi)\n",
    "        new_score = score(motifs, profile)\n",
    "        if new_score < best_score:\n",
    "            best_score = new_score\n",
    "            best_motifs = motifs\n",
    "    return best_motifs\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
